LibHR/Memory/TMPL/field_alloc.c.tmpl

```diff
@@ -0,0 +1,189 @@
+/***************************************************************************\
+* Copyright (c) 2008, 2022, Claudio Pica, Sofie Martins                     *   
+* All rights reserved.                                                      * 
+\***************************************************************************/
+
+// TODO: Deallocate sendbuffers
+ 
+/**
+ * @file field_alloc.c.tmpl
+ * @brief Generics for allocation and free functions for all fields defined in spinor_field.h
+ */
+
+#include "memory.h"
+#include "libhr_core.h"
+#include "Utils/generics.h"
+
+#if !defined _FIELD_NAME
+#error Missing _FIELD_NAME in memory.c
+#endif
+#if !defined _FIELD_TYPE
+#error Missing _FIELD_TYPE in memory.c
+#endif
+#if !defined _SITE_TYPE
+#error Missing _SITE_TYPE in memory.c
+#endif
+#if !defined _FIELD_DIM
+#error Missing _FIELD_DIM in memory.c
+#endif
+#if !defined _IS_SPINOR_LIKE
+#error Missing _IS_SPINOR_LIKE in memory.c
+#endif
+
+#define _F_ARGS(_name,_args) _FIELD_TYPE* _F_NAME(_name,_FIELD_NAME) _args
+
+// This is necessary, because spinor-like fields have an additional argument 
+// in the allocation function
+#if _IS_SPINOR_LIKE==0
+    #define _n 1
+    #define _GEOM_TYPE gauge
+    #define _ALLOCATE _F_ARGS(alloc_,(geometry_descriptor *type))
+#else
+    #define _n n
+    #define _GEOM_TYPE spinor
+    #define _ALLOCATE _F_ARGS(alloc_,(unsigned int n, geometry_descriptor *type))
+#endif
+
+#define _DECLARE(_name,_args) _FUNC(void, _name, _FIELD_NAME, _args)
+#define _DECLARE_STATIC(_name,_args) _FUNC(static void, _name, _FIELD_NAME, _args)
+#define _CALL(_name) _F_NAME(_name,_FIELD_NAME)
+
+_DECLARE_STATIC(allocate_field_struct_,(unsigned int n, _FIELD_TYPE** f, geometry_descriptor *type)) {
+    *f = amalloc(n * sizeof(_FIELD_TYPE), ALIGN);
+    error(f == NULL, 1, __func__, "Could not allocate memory space for field structure");
+    for (int i = 0; i < n; ++i) { (*f)[i].type = type; }
+}
+
+_DECLARE_STATIC(allocate_cpu_field_data_,(unsigned int n, _FIELD_TYPE *f, geometry_descriptor *type)) {
+    if (alloc_mem_t & CPU_MEM) {
+        int bytes_per_site = sizeof(_SITE_TYPE);
+        int number_of_sites = n * _FIELD_DIM * _NUMBER_OF_SITES(_GEOM_TYPE);
+        int field_size  = bytes_per_site * number_of_sites;
```

Comment:
these int variables overflow for large lattices. can you declare them as size_t ?
Same below for GPU code